### DISCUSSION TITLE
Skeleton Roadmap

### DISCUSSION THREAD

**Discussion by endigo9740** [skeletonlabs/skeleton#1367] [Announcements]
> **Note** This is a living document and may update periodically

View past Roadmaps:

- [v1 Roadmap](https://github.com/skeletonlabs/skeleton/discussions/503)
- [v2 Roadmap](https://github.com/skeletonlabs/skeleton/discussions/1961)

---

Hey everyone, Chris here from the Skeleton crew. We've had a lot of questions about the future plans for Skeleton so I wanted to pull back the curtain and share what we've been cooking up.

Please note that the Skeleton core team is comprised of four contributor role members (Nik, Rhys, Adrian, and myself) and four moderators (Dom, Luke, Brent, and Sebas). After the v1.0 launch the full team and I met to discuss the future of the project and define our long term goals. While the team is responsible for organizing these ideas, most suggestions has come either directly or indirectly from feedback and suggestions provided by you (our community).

The information below represents a high level overview and future plans. Nothing here is set in stone, so expect revisions. However, this is meant to provide our general direction for the project. We do very much welcome your feedback in the comment section below!

> **Note** items marked as `PROPOSAL` are still under review

---

# Skeleton v3+

- [ ] Migrate features to dedicated workspace projects within the monorepo, including:
  - [ ] The theme generator - standalone site + https://github.com/skeletonlabs/skeleton/issues/1142
  - [ ] https://github.com/skeletonlabs/skeleton/issues/1916
  - [ ] The codeblock component - standalone package, w/ Highlight.js peer dependency
  - [ ] PROPOSAL: remove [LocalStoreStore](https://github.com/joshnuss/svelte-local-storage-store/releases) from the core library (note the change in v0.5.0 we don't yet support)
- [ ] Documentation Site
  - [ ] [Performance Optimization](https://github.com/skeletonlabs/skeleton/issues/1149)
  - [ ] Versioning - at least for major releases
- [ ] Tailwind Updates
  - [ ] https://github.com/skeletonlabs/skeleton/issues/1238
  - [x] PROPOSAL: https://github.com/skeletonlabs/skeleton/issues/1286
  - [ ] PROPOSAL: migrate Element/Component styles to utilize container queries
  - [ ] PROPOSAL: Rethinking Dark Mode Strategy (describe in detail below)
- [ ] Accessibility
    - [ ] https://github.com/skeletonlabs/skeleton/issues/1708
    - [ ] Improved theme accessibility, especially in regards to high/low contrast designs
    - [ ] Improved keyboard interaction to meet ARIA APG Guidelines
    - [ ] Audit and improve ARIA roles and attributes to meet ARIA APG Guidelines
- [ ] Feature Update
  - [ ] (TBD as we work through our [backlog](https://github.com/skeletonlabs/skeleton/issues?q=is%3Aopen+is%3Aissue+label%3A%22feature+request%22))
- [x] Support for the [Svelte v4 release](https://github.com/sveltejs/svelte/releases)
- [ ] Refactor `focus-trap` to include a new dedicated `strict` argument ([reference](https://github.com/skeletonlabs/skeleton/pull/1614#issuecomment-1572635597))
- [ ] Consider removing LocalStorageStore from the core library, CLI installed dependency
- [ ] Proposed Breaking Changes
    - [ ] Standardize on `region` versus `slot` naming for properties.

---
 
# Long-Term Concepts and Proposals

> **Warning** the following are merely proposals, these may or may not move forward.

Please note the following represents some "big picture" ideas we have for the far flung future of Skeleton. We reserve the right to change the direction and/or discard any of the following concepts at any time. In short, temper your expectations. However, we do wish to share our big ideas early to help garner support and contributors and help guide the overall direction the project is heading. If these features move forward, expect to see them as major breaking change releases, such as v2, v3, etc.

## Rethinking Dark Mode Strategy

I've written a detailed post on this subject, which can be found in the link below. In short, we have some flaws in how light and dark mode is handled in relation to features such as design tokens. Given the scale and scope of this proposal, we do highly encourage feedback from you on this. Additionally, upcoming changes in next major version of Tailwind may influence how we approach this as well.

- https://github.com/skeletonlabs/skeleton/discussions/951

## Headless Components

This is a hard concept to explain with words alone, so I've created diagram below to help visualize this proposal:

![Screenshot 2023-04-24 at 1 55 23 PM](https://user-images.githubusercontent.com/1509726/234089849-de1b7dc6-b380-4ad3-9f77-5501603c679b.png)

Currently Skeleton components are a tightly coupled set of HTML templates, TS/JS logic, and Tailwind CSS utility classes. This provides a fully featured component that comes pre-styled, while still adapting to your current theme. However, there are benefits to splitting the components between what are commonly referred to as "headless components" and our unique approach to Tailwind Element styling. Here's a few potential use cases with the split pattern:

- Headless components can be used standalone with completely custom styles implemented
- Tailwind Element styles can be used standalone for custom built components
- Headless components + Tailwind Elements can be paired together to operate similar to how things work today

## Going Framework Agnostic

We're often asked why we didn't name the library "Sveleton" or something more Svelte-specific, and the reason for that is simple: we've always anticipated the library might one day go framework agnostic. We love Svelte and SvelteKit and imagine it will continue to be the tip of the spear for us, leading all feature development for the foreseeable future. However, we often hear it said that Skeleton has brought a lot of new ideas and innovations to the UI library space (such as our adapting themes and use of design tokens). We think there's potential to bring these benefits to ALL frameworks, including those not yet created. The headless component + element styling approach is one of the primary features that may make this possible.

Again, here's another diagram to visualize how this might work:

![Screenshot 2023-04-24 at 2 03 44 PM](https://user-images.githubusercontent.com/1509726/234091484-963c9d1f-da32-4d1f-bf71-ecd38ebd6dbc.png)

By separating the visual presentation (Tailwind Element styles) from the core component itself, it makes it much easier to port each component to a new Javascript Framework, such as React, Vue, Angular, etc. Tailwind and styling would then be the shared commonality between each. While not a trivial task by any stretch of the imagination, especially given how each framework operates differently (ex: Stores vs Signals vs rxJS Observables). We do think this may be technically possible

That said, this would be a tremendous amount of effort, likely taking multiple years to fully implement. We would also need the assistance of one or more experts in each framework we wished to support. If you think you might be capable of helping with something like this, and understand the long term commitment required to make this happen, then please reach out to me (@endigo9740) on the Skeleton Discord or email me at chris@skeleton.dev. I'd love to meet and review this in more detail!

# Sharing Feedback

If you have questions or feedback about anything above, please consider leaving a comment in the section below. Thanks!

**Comment by saturnonearth** [comment]
Awesome Roadmap!

Some brief thoughts after reading it.

> Headless components

With the radix-svelte port being worked on, and gainings some traction, this is definitely an interesting one.

I think separating the styles from the components themselves could be a great idea. You don't want to have to wrestle with a framework, like the good ole' days of Bootstrap 2. But at the same time, it can good to _stay_ a a somewhat opinionated framework. Having element styles be separate would likely make the boilerplate a bit more verbose, causing simple items to look more complicated than they are (we all know how nasty/lengthy Tailwind classes can get). 

Honestly I am stuck on this one - because on one hand, having to override a padding that a button has by default when next to another element can be frustrating, and on the other hand, those little CSS adjustments can be real time savers.

Bootstrap (not that Skeleton is Bootstrap at all) was popular for a reason; it was hated for that same reason.

I wonder if having some sort of wizard on the site to adjust common things like padding and margins throughout forms, or gap sizes throughout the theme, would be a better plan. What are your thoughts? What complaints do you hear most from the Skeleton community regarding styles?

> Going Framework Agnostic

Personally I think this is a mistake. Of course I am a little biased since I use Svelte but I think trying to support multiple frameworks would slow down development on all fronts.

There are already so many good frameworks, both style opinionated ([Chakra UI)](https://chakra-ui.com/), and headless ([Radix](https://www.radix-ui.com/))...and even ones built on headless ([shadn-ui](https://ui.shadcn.com/)).

React is such a rich eco-system, and has had time to develop, that I think even splitting any resources away from the Svelte community would be an unfortunate step. Sure, I think opening Skeleton to other frameworks would probably bring more interest, and could in turn even help out the Svelte side inadvertently, but I would rather not bet on it.

**Reply by endigo9740** [reply]
Thanks @saturnonearth, you brought up a couple great points. I was trying to keep the explanations above slim and simple so not to overwhelm, but I have a lot more in mind for all these proposals. Too much to fit here.

> Headless components

The way I'd imagine this working is we would build on our system for feeding utility classes as props for components. Instead of feeding primative utility classes (ex: bg-green-500) to a `background` prop, you could feed an entire Skeleton Tailwind Element class to the prop, like `<Table element="table">`, where `.table` represents the parent element class. This means two things:

1. These props could have a default, so by default a Table component uses Skeleton's `.table` element class. Meaning the Skelton styles are on by default.
2. These could easily be exchanged with your own class, like `<table element="myTableClass">`. The component will revert to a naked/headless state until your custom `.myTableClass` styles are provided.

It's a lot like the current prop system, just on crack! Another cool benefit of this is the classes provided by the user don't have to be Tailwind. So traditional CSS styling is available for those that shy away from Tailwind. Though I'm not sure if we can drop the Tailwind dependency altogether, all our "in-house" styles will still utilize Tailwind.

> I wonder if having some sort of wizard on the site to adjust common things like padding and margins throughout forms, or gap sizes throughout the theme, would be a better plan. What are your thoughts? What complaints do you hear most from the Skeleton community regarding styles?

I actually have always intended to implement margins/padding into theme settings, then build design tokens around them. Then you get the nice standardization approach that tokens provide. Meaning you may no longer be required to set a padding for every card instance. Though margin/padding is more nuanced than, say, border radius. We might need canned presets like `sm, md, lg` margin/padding.

> Personally I think this is a mistake. Of course I am a little biased since I use Svelte but I think trying to support multiple frameworks would slow down development on all fronts.

It very well could be. You're not the first to express this concern. There's a time and maintenance cost for every framework we support. Without additional full time staff this falls apart, as I can't be the expert leading the charge for every framework. It's too much effort. But I know library ports are common (ex: Svelte-Radix within the last couple weeks). So in some ways this in inevitable. Heck, I already know of one guy porting Skeleton to Vue in his free time, for example. I think we can lay a lot of the foundation for this with other changes we know we want to implement though, then only worry about expanding if/when it makes sense.

Funny enough I'm not really interested in expanding to React. I'd rather look at the "new kids on the block" like Solid and Quik and whatever comes next.

**Reply by bartduisters** [reply]
> Heck, I already know of one guy porting Skeleton to Vue in his free time, for example.

I wonder what maniac spends his time doing such a thing.

**Reply by saturnonearth** [reply]
> where .table represents the parent element class. This means two things....It's a lot like the current prop system, just on crack!

Ah  I understand now. This sounds great!

> I actually have always intended to implement margins/padding into theme settings 

Cool!

**Reply by FlipOneForEtx** [reply]
Maybe this: Core developers focus on Svelte (there's still too much to do), let others port Skeleton to other frameworks.

**Reply by baseplate-admin** [reply]
> Maybe this: Core developers focus on Svelte (there's still too much to do), let others port Skeleton to other frameworks.

Yep. Like how radix is being ported to svelte by another dev

**Comment by KoljaL** [comment]
I really celebrate the headless plans!

**Comment by brucewayyn** [comment]
Headless Skeleton is what I have suggested since the earliest days of Skeleton! In fact, I have argued about it on multiple occasions with Chris on Discord (as Shoob).

**Comment by multiwebinc** [comment]
I'm a React developer and I was really drawn by the design tokens aspect of Skeleton. I actually started looking just today to see if it would be possible to port that aspect over to a Next.js app I'm working on. So I would for sure love if the design tokens were framework agnostic.

As for developing actual components in all of the other frameworks, I don't think it's really worth it. I know, at least with React, there are already other headless component libraries, so you would be reinventing the wheel. Just being able to style them with Skeleton's design tokens would be amazing.

**Reply by endigo9740** [reply]
Yeah @multiwebinc would be glad to share how some of that works. It's definitely not Svelte specific.

And I don't see us supporting all frameworks, but I think a lot of the changes we have planned will just make it possible. So we'll see if if there's anything we can do with that. But I definitely see that a year or two away. Nothing in the immediate future.

**Reply by multiwebinc** [reply]
@endigo9740 I actually followed the manual instructions on https://www.skeleton.dev/docs/get-started and I was able to get it working with Next.js, Radix and shadcn-ui components (although I am likely going to have to overwrite all of the styles from them or maybe just use straight Radix):

1. Install Next.js and Radix. I did this by installing `shadcn/ui` according to https://ui.shadcn.com/docs/installation
2. Install dependencies (hopefully I didn't miss any):
```
npm i @skeletonlabs/skeleton --save-dev
npm i postcss-import --save-dev
npm i @tailwindcss/forms --save-dev
npm i @tailwindcss/typography --save-dev
```
3. Change your `postcss.config.js` file to:
```js
module.exports = {
  plugins: {
    "postcss-import": {},
    tailwindcss: {},
    autoprefixer: {},
  },
}
```
4. Combine the `tailwind.config.js` file with the one from Skeleton (https://github.com/skeletonlabs/skeleton/blob/dev/tailwind.config.cjs).
5. Add the following to your layout file:

```js
import "@skeletonlabs/skeleton/themes/theme-skeleton.css"
import "@skeletonlabs/skeleton/styles/all.css"
```
6. (optional) Remove the import of `@/styles/globals.css` (since you don't need it anymore, or instead you could just start with a fresh file)
7. (optional) Add `data-theme="skeleton"` to your `<body>` tag in the layout.

You should now be able to use Skeleton classes:

```jsx
<button type="button" className="btn-icon variant-soft-primary">
  (icon)
</button>
<button type="button" className="btn variant-soft-primary">
  Button
</button>
<button type="button" className="btn variant-soft-primary">
  <span>(icon)</span>
  <span>Button</span>
</button>
```

![image](https://user-images.githubusercontent.com/901732/236596183-fc424ed6-d9c5-435f-99b3-17f648e537e2.png)

**Reply by endigo9740** [reply]
Awesome work @multiwebinc! I meant to mention that everything but our components and utility features will work cross-framework. Tailwind is really the only requirement there. Not a bad way to dip your toes in for sure!

**Comment by Hugos68** [comment]
I really am glad this library might go headless in the future. This would allow other styling frameworks/libraries to be paired with Skeleton and not limit to only tailwind, this would also allow for existing projects to use Skeleton purely for logic (which is usually the hard part of any UI pattern) because some projects might already use something else.
I'm super glad to see this.
Decoupling = Good.

**Comment by endigo9740** [comment]
Just FYI everyone, we'll be making some changes for Skeleton v3. See the RFC and proposal to learn what's incoming:
https://github.com/skeletonlabs/skeleton/discussions/2375

And you can track progress via a dedicated milestone here:
https://github.com/skeletonlabs/skeleton/milestone/5
